### PR TITLE
[13.0][FW] account_multicurrency_revaluation: forward port PR#149

### PR DESCRIPTION
**Pull request by JordiMForgeFlow**
Superseeds #183

### CODE REVIEW

**Comment by JordiMForgeFlow** [comment]
Hi @alexey-pelykh, @hailangvn and @bunhin

Any chance you can help me reviewing the changes? :)

**Review by hailangvn — APPROVED** [review]
LGTM. Thanks for the work.

**Review by alexey-pelykh — APPROVED** [review]
Code LGTM

**Comment by GuillemCForgeFlow** [comment]
Hi @hailangvn , may you be able to review the changes I've added? :smile:

File: account_multicurrency_revaluation/model/res_company.py [hunk 1]
```diff
@@ -77,4 +77,5 @@ class ResCompany(models.Model):
         string="Reversable Revaluations",
         help="Revaluations entries will be created " 'as "To Be Reversed".',
         default=True,
+        readonly=False,
```
Comment (alexey-pelykh) [inline account_multicurrency_revaluation/model/res_company.py:83]: why this is needed?

Comment (GuillemCForgeFlow) [reply account_multicurrency_revaluation/model/res_company.py:83]: Just to make sure that the field appears editable on the view. I could remove it and only add it to the field of the res_config file.

Comment (alexey-pelykh) [reply account_multicurrency_revaluation/model/res_company.py:83]: Yes, readonly=False should be done on related field, not here

Comment (GuillemCForgeFlow) [reply account_multicurrency_revaluation/model/res_company.py:83]: Great, I have applied the changes!

**Comment by alexey-pelykh** [comment]
@GuillemCForgeFlow please rebase one more time to pickup new CI

File: account_multicurrency_revaluation/model/account.py [hunk 2]
```diff
@@ -54,83 +78,109 @@ def _onchange_user_type_id(self):
             if rec.user_type_id.id in revaluation_accounts:
                 rec.currency_revaluation = True
 
-    def _revaluation_query(self, revaluation_date):
-
+    def _revaluation_query(self, start_date, revaluation_date):
```
Comment (hailangvn) [inline account_multicurrency_revaluation/model/account.py]: IMHO we can enhance compatibility by using different method signature.
```suggestion
    def _revaluation_query(self, revaluation_date, start_date=None):
```

File: account_multicurrency_revaluation/model/account.py [hunk 3]
```diff
@@ -54,83 +78,109 @@ def _onchange_user_type_id(self):
...
+            params.insert(4, start_date)
+
         return query, params
 
-    def compute_revaluations(self, revaluation_date):
-        accounts = {}
-        # compute for each account the balance/debit/credit from the move lines
-        query, params = self._revaluation_query(revaluation_date)
+    def compute_revaluations(self, start_date, revaluation_date):
+        query, params = self._revaluation_query(start_date, revaluation_date)
```
Comment (hailangvn) [inline account_multicurrency_revaluation/model/account.py]: Especially for public method like this.
```suggestion
    def compute_revaluations(self, revaluation_date, start_date=None):
        query, params = self._revaluation_query(revaluation_date, start_date)
```

File: account_multicurrency_revaluation/wizard/print_currency_unrealized_report.py [hunk 4]
```diff
@@ -13,26 +15,61 @@ class UnrealizedCurrencyReportPrinter(models.TransientModel):
     account_ids = fields.Many2many(
         "account.account",
         string="Accounts",
-        domain="[('currency_revaluation', '=', True)]",
+        domain=[("currency_revaluation", "=", True)],
         default=lambda self: self._default_account_ids(),
     )
+    start_date = fields.Date(
+        string="Start Date",
+        help="The report will print from this Date, all the revaluated entries"
```
Comment (hailangvn) [inline account_multicurrency_revaluation/wizard/print_currency_unrealized_report.py:23]: I think space is needed at end of this line (or at beginning of next line).
```suggestion
        help="The report will print from this Date, all the revaluated entries "
```

File: account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py [hunk 5]
```diff
@@ -310,83 +277,102 @@ def revaluate_currency(self):
...
                     "Please check 'Allow Currency Revaluation' "
                     "for at least one account in account form."
                 )
             )
 
-        # Get balance sums
-        account_sums = account_ids.compute_revaluations(self.revaluation_date)
+        revaluations = account_ids.compute_revaluations(
+            self.start_date, self.revaluation_date
+        )
```
Comment (hailangvn) [inline account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py]:
```suggestion
        revaluations = account_ids.compute_revaluations(
            self.revaluation_date, self.start_date
        )
```

**Review by hailangvn — COMMENTED** [review]
Thanks for the work. It looks good except some minor points to help improving experience of users and developers. And as we are on this, please consider to update below help string and its translations.

* [account_multicurrency_revaluation/model/res_company.py lines 84-85](https://github.com/OCA/account-closing/pull/200/files#diff-2c16930dd0dbc7275aa3510f3e0dce93fcf838e476db03cfa7db71ac594bd5f8R84-R85) add a space at end/beginning of a line.

```
        help="If marked, the entries created in the process will be "
        "posted automatically.",
```

**Comment by GuillemCForgeFlow** [comment]
Hi @hailangvn , you're right! I have added you're suggestions :+1:

File: account_multicurrency_revaluation/model/res_company.py [hunk 6]
```diff
@@ -76,13 +76,13 @@ class ResCompany(models.Model):
...
         help="If marked, the entries created in the process will be flaged "
-        "to be able to filter them out later.",
+        "to be able to filter them out later. ",
         default=True,
     )
     auto_post_entries = fields.Boolean(
         string="Auto Post Created Entries",
         help="If marked, the entries created in the process will be"
-        "posted automatically.",
+        "posted automatically. ",
```
Comment (hailangvn) [inline account_multicurrency_revaluation/model/res_company.py]: The space is needed to separate `be` and `posted` so below is the suggested change.
```suggestion
        " posted automatically.",
```

File: account_multicurrency_revaluation/wizard/print_currency_unrealized_report.py [hunk 7]
```diff
@@ -21,12 +21,13 @@ class UnrealizedCurrencyReportPrinter(models.TransientModel):
     start_date = fields.Date(
         string="Start Date",
         help="The report will print from this Date, all the revaluated entries"
-        "created from this date. The default value will be the first day of the month.",
+        "created from this date. The default value will be the first day of the month."
+        " ",
```
Comment (hailangvn) [inline account_multicurrency_revaluation/wizard/print_currency_unrealized_report.py]: The space is needed to separate `entries` with `created` so below is the suggested change.
```suggestion
        " created from this date. The default value will be the first day of the month."
```

File: account_multicurrency_revaluation/wizard/print_currency_unrealized_report.py [hunk 8]
```diff
@@ -21,12 +21,13 @@ class UnrealizedCurrencyReportPrinter(models.TransientModel):
...
         default=lambda self: self._default_start_date(),
     )
     end_date = fields.Date(
         string="End Date",
-        help="The report will print till this Date. The default value will be today.",
+        help="The report will print till this Date. The default value will be today. ",
```
Comment (hailangvn) [inline account_multicurrency_revaluation/wizard/print_currency_unrealized_report.py]: This one only has one line so no need space at end.
```suggestion
        help="The report will print till this Date. The default value will be today.",
```

File: account_multicurrency_revaluation/model/res_company.py [hunk 9]
```diff
@@ -76,13 +76,13 @@ class ResCompany(models.Model):
     reversable_revaluations = fields.Boolean(
         string="Flag Revaluation Entries",
         help="If marked, the entries created in the process will be flaged "
-        "to be able to filter them out later.",
+        "to be able to filter them out later. ",
```
Comment (hailangvn) [inline account_multicurrency_revaluation/model/res_company.py]: The previous line has space at end already so no need to change this line.
```suggestion
        "to be able to filter them out later.",
```

File: account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py [hunk 10]
```diff
@@ -56,7 +56,7 @@ def _get_default_revaluation_account_ids(self):
         comodel_name="account.journal",
         string="Journal",
         domain=[("type", "=", "general")],
-        help="You can set the default journal in company settings.",
+        help="You can set the default journal in company settings. ",
```
Comment (hailangvn) [inline account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py]: This one only has one line so no need space.
```suggestion
        help="You can set the default journal in company settings.",
```

File: account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py [hunk 11]
```diff
@@ -72,7 +72,7 @@ def _get_default_revaluation_account_ids(self):
     revaluation_account_ids = fields.Many2many(
         comodel_name="account.account",
         string="Revaluation Accounts",
-        help="Accounts that will be revaluated",
+        help="Accounts that will be revaluated. ",
```
Comment (hailangvn) [inline account_multicurrency_revaluation/wizard/wizard_currency_revaluation.py]: This one only has one line so no need space.
```suggestion
        help="Accounts that will be revaluated.",
```

**Review by hailangvn — APPROVED** [review]
Appreciate your effort for this. Sorry that I was not specific enough in previous comments. So the minimum changes are the two spaces to separate words. Other changes are just nitpicking on the way.

**Comment by GuillemCForgeFlow** [comment]
Hi @hailangvn , now I got you! I've applied the suggestions :+1:

**Review by hailangvn — APPROVED** [review]
LGTM. Thank you.

**Review by MiquelRForgeFlow — APPROVED** [review]
I don't know if it's okey having that many commits instead of squashing... but LGTM

**Review by GuillemCForgeFlow — APPROVED** [review]

**Review by AaronHForgeFlow — APPROVED** [review]
LGTM

**Pull request merged**